R/utils.R

```diff
@@ -4,36 +4,77 @@
 #' @importFrom stats median rnorm
 NULL
 
-#' @title FUNCTION_TITLE
-#' @description FUNCTION_DESCRIPTION
-#' @param files PARAM_DESCRIPTION
-#' @param target_dir PARAM_DESCRIPTION
-#' @param ... PARAM_DESCRIPTION
-#' @return OUTPUT_DESCRIPTION
+#' Read in a stan function file as a character string
+#'
+#' @inheritParams expose_stan_fns
+#' @return A character string in the of stan functions.
 #' @family utils
-#' @export
 #' @importFrom purrr map_chr
-#' @importFrom rstan expose_stan_functions stanc
-expose_stan_fns <- function(files, target_dir, ...) {
+stan_fns_as_string <- function(files, target_dir) {
   functions <- paste0(
     "\n functions{ \n",
-    paste(purrr::map_chr(
-      files,
-      ~ paste(readLines(file.path(target_dir, .)), collapse = "\n")
-    ),
-    collapse = "\n"
-    ),
+      paste(purrr::map_chr(
+        files,
+        ~ paste(readLines(file.path(target_dir, .)), collapse = "\n")
+      ),
+      collapse = "\n"
+      ),
     "\n }"
   )
+  return(functions)
+}
 
-  functions <- gsub("array\\[\\] ([a-z]+) ([a-z_]+)", "\\1[] \\2", functions)
+#' Convert Cmdstan to Rstan
+#'
+#' @param functions A character string of stan functions as produced using
+#' [stan_fns_as_string()].
+#'
+#' @return A character string of stan functions converted for use in `rstan`.
+#' @family utils
+convert_cmdstan_to_rstan <- function(functions) {
+  # replace bars in CDF with commas
+  functions <- gsub("_cdf\\(([^ ]+) *\\|([^)]+)\\)", "_cdf(\\1,\\2)", functions)
+  # replace lupmf with lpmf
+  functions <- gsub("_lupmf", "_lpmf", functions)
+  # replace array syntax
+  #   case 1: array[] real x -> real[] x
   functions <- gsub(
-    "array\\[,\\] ([a-z]+) ([a-z_]+)", "\\1[,] \\2", functions
+    "array\\[(,?)\\] (.*) ([a-z_]+)", "\\2[\\1] \\3", functions
   )
+  #   case 2: array[n] real x -> real x[n]
   functions <- gsub(
-    "array\\[([a-z]+)\\] ([a-z]+) ([a-z_]+)", "\\2 \\3[\\1]", functions
+    "array\\[(.*)\\] (.*) ([a-z_]+)", "\\2 \\3[\\1]", functions
```

Comment:
I think the first one is slightly risky as it will go greedily across lines. I'd suggest to replace `array\\[(.*)\\]` with `array\\[([^]]*)\\]` to make sure it stops at the first closing bracket.

Reply:
If aiming to be as general as possible might also want to ensure it works with any amount of any whitespace by changing to something like
`"array\\[([^]]*)\\]\\s+(.*)\\s+([a-z_]+)", "\\2 \\3[\\1]"`

Reply:
hahaha I hate regex.

Reply:
https://github.com/epiforecasts/epinowcast/pull/93